ddl/ddl.go

```diff
@@ -642,45 +646,106 @@ func (d *ddl) prepareWorkers4legacyDDL() {
+
+	// Start some background routine to manage TiFlash replica.
+	d.wg.Run(d.PollTiFlashRoutine)
 
-		err := d.ownerManager.CampaignOwner()
+	return nil
+}
+
+// EnableDDL enable this node to execute ddl.
+// Since ownerManager.CampaignOwner will start a new goroutine to run ownerManager.campaignLoop,
+// we should make sure that before invoking EnableDDL(), ddl is DISABLE.
+func (d *ddl) EnableDDL() error {
+	d.ownerManager.SetBeOwnerHook(func() {
+		var err error
+		d.ddlSeqNumMu.seqNum, err = d.GetNextDDLSeqNum()
 		if err != nil {
-			return errors.Trace(err)
+			logutil.BgLogger().Error("error when getting the ddl history count", zap.Error(err))
 		}
+	})
 
-		d.delRangeMgr = d.newDeleteRangeManager(ctxPool == nil)
+	err := d.ownerManager.CampaignOwner()
+	if err != nil {
+		return errors.Trace(err)
+	}
 
-		d.prepareWorkers4ConcurrencyDDL()
-		d.prepareWorkers4legacyDDL()
+	d.delRangeMgr = d.newDeleteRangeManager(d.sessPool.resPool == nil)
 
-		go d.schemaSyncer.StartCleanWork()
-		if config.TableLockEnabled() {
-			d.wg.Add(1)
-			go d.startCleanDeadTableLock()
-		}
-		metrics.DDLCounter.WithLabelValues(metrics.StartCleanWork).Inc()
+	d.prepareWorkers4ConcurrencyDDL()
+	d.prepareWorkers4legacyDDL()
+
+	go d.schemaSyncer.StartCleanWork()
+	if config.TableLockEnabled() {
+		d.wg.Add(1)
+		go d.startCleanDeadTableLock()
 	}
+	metrics.DDLCounter.WithLabelValues(metrics.StartCleanWork).Inc()
+	return errors.Trace(err)
+}
 
-	variable.RegisterStatistics(d)
+// DisableDDL disable this node to execute ddl.
+// We should make sure that before invoking DisableDDL(), ddl is ENABLE.
+func (d *ddl) DisableDDL() error {
+	if d.ownerManager.IsOwner() {
+		// If there is only one node, we should NOT disable ddl.
+		serverInfo, err := infosync.GetAllServerInfo(d.ctx)
+		if err != nil {
+			logutil.BgLogger().Error("[ddl] error when GetAllServerInfo", zap.Error(err))
+			return err
+		}
+		if len(serverInfo) <= 1 {
+			return dbterror.ErrDDLSetting
+		}
 
-	metrics.DDLCounter.WithLabelValues(metrics.CreateDDLInstance).Inc()
+		// FIXME: if possible, when this node is the only node with DDL, ths setting of DisableDDL should fail.
 
-	// Start some background routine to manage TiFlash replica.
-	d.wg.Run(d.PollTiFlashRoutine)
+		// lets the owner start a new election.
+		err = d.ownerManager.ResignOwner(d.ctx)
+
+		if err != nil {
+			logutil.BgLogger().Error("[ddl] error when ResignOwner", zap.Error(err))
+			return err
+		}
+	}
+
+	d.cancel()
+	d.wg.Wait()
+	d.schemaSyncer.Close()
+	if d.reorgWorkerPool != nil {
+		d.reorgWorkerPool.close()
+	}
+	if d.generalDDLWorkerPool != nil {
+		d.generalDDLWorkerPool.close()
+	}
+
+	for _, worker := range d.workers {
+		worker.Close()
+	}
+	// d.delRangeMgr using sessions from d.sessPool.
+	// Put it before d.sessPool.close to reduce the time spent by d.sessPool.close.
+	if d.delRangeMgr != nil {
+		d.delRangeMgr.clear()
+	}
+	// disable campaign by interrupting campaignLoop
+	d.ownerManager.CampaignCancel()
```

Comment:
`CampaignCancel` will not confirm the campaign loop is over. Need confirmation from the waitgroup. Refer `ownerManager.Cancel()`.